src/System.Net.Security/src/System/Net/Security/SslStreamInternal.cs

```diff
@@ -270,137 +236,132 @@ private AsyncProtocolRequest GetOrCreateProtocolRequest(ref AsyncProtocolRequest
             return request;
         }
 
-        //
-        // Sync write method.
-        //
-        private void ProcessWrite(byte[] buffer, int offset, int count, LazyAsyncResult asyncResult)
+        private Task WriteAsyncInternal<TWriteAdapter>(TWriteAdapter writeAdapter, byte[] buffer, int offset, int count)
+            where TWriteAdapter : struct, ISslWriteAdapter
         {
             _sslState.CheckThrow(authSuccessCheck: true, shutdownCheck: true);
-            ValidateParameters(buffer, offset, count);
 
-            if (Interlocked.Exchange(ref _nestedWrite, 1) == 1)
+            if (count == 0 && !SslStreamPal.CanEncryptEmptyMessage)
             {
-                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, "Write", "write"));
+                // If it's an empty message and the PAL doesn't support that, we're done.
+                return Task.CompletedTask;
             }
 
-            // If this is an async operation, get the AsyncProtocolRequest to use.
-            // We do this only after we verify we're the sole write operation in flight.
-            AsyncProtocolRequest asyncRequest = GetOrCreateProtocolRequest(ref _writeProtocolRequest, asyncResult);
-
-            bool failed = false;
+            if (Interlocked.Exchange(ref _nestedWrite, 1) == 1)
+            {
+                throw new NotSupportedException(SR.Format(SR.net_io_invalidnestedcall, nameof(WriteAsync), "write"));
+            }
 
-            try
+            if (count <= _sslState.MaxDataSize)
             {
-                StartWriting(buffer, offset, count, asyncRequest);
+                //Single write
+                Task t = WriteSingleChunk(writeAdapter, buffer, offset, count);
+                if (t.IsCompletedSuccessfully)
+                {
+                    _nestedWrite = 0;
+                    return t;
+                }
+                else
+                {
+                    return ExitWriteAsync(t);
+                }
             }
-            catch (Exception e)
+            else
             {
-                _sslState.FinishWrite();
+                return ExitWriteAsync(WriteAsyncChunked(writeAdapter, buffer, offset, count));
+            }
```

Comment:
Nit: this could be done later if desired, but we could reorganize this slightly to avoid multiple calls to ExitWriteAsync, e.g.
```C#
Task t;
if (count < _sslState.MaxDataSize)
{
    // Single write
    t = WriteSingleChunk(writeAdapter, buffer, offset, count);
    if (t.IsCompletedSuccessfully)
    {
        _nestedWrite = 0;
        return t;
    }
}
else
{
    t = WriteAsyncChunked(writeAdapter, buffer, offset, count);
}
return ExitWriteAsync(t);
```

Reply:
Changed

Reply:
And actually, looks like it could be simplified even further to the following (unless I'm missing something):
```C#
Task t = count < _sslState.MaxDataSize ?
    WriteSingleChunk(writeAdapter, buffer, offset, count) :
    WriteAsyncChunked(writeAdapter, buffer, offset, count);
if (t.IsCompletedSuccessfully)
{
    _nestedWrite = 0;
    return t;
}
return ExitWriteAsync(t);
```

Reply:
yeah I have just been looking at that, after the adjustment it becomes clear that we don't need to call the ExitWriteAsync for every call to WriteAsyncChunked

Changed.